src/sorcha_community_utils/activity/lsst_comet/model.py

```diff
@@ -0,0 +1,129 @@
+
+    k : float
+        Activity power-law slope with heliocentric distance: ``rh^k``.
+
+    q : float, required if ``afrho_q`` is provided
+        Perihelion distance for Afρ normalization.
+
+    Phi_c : function, optional
+        One-parameter function that takes phase angle in degrees and
+        returns the phase function for comae.  It is assumed that
+        Phi(0) = 1.0.  Default is to use the Halley-Marcus phase
+        function of Schleicher & Bair 2011.
+
+    Phi_n : function, optional
+        Same as ``Phi_c`` but for nuclei.  Default is 0.04 mag/deg.
+
+    """
+
+    # Willmer 2018, ApJS 236, 47
+    mv_sun = -26.76  # Vega mag
+    m_sun = {"u": -25.30, "g": -26.52, "r": -26.93, "i": -27.05, "z": -27.07, "y": -27.07}  # AB mag
+
+    def __init__(self, R, k=-2, afrho1=1500, **kwargs):
+        self.R = R
+        self.k = k
+        self.afrho1 = afrho1
+
+        self.Phi_c = kwargs.get("Phi_c", phase_HalleyMarcus)
+
+    def _get_value(self, d, keys, unit):
+        v = None
+        for k in keys[::-1]:
+            try:
+                # this works with sbpy Ephem objects
+                v = d[k]
+                break
+            except KeyError:
+                pass
+
+        if v is None:
+            raise IndexError
+        if u:
+            v = u.Quantity(v, unit).value
+        return v
+
+    def _normalize_geom(self, geom):
+        # allows for sbpy Ephem objects, LSST MAF ssObs, and plain
+        # dictionaries
+        return {
+            "rh": self._get_value(geom, ("rh", "helio_dist"), "au"),
+            "delta": self._get_value(geom, ("delta", "geo_dist"), "au"),
+            "phase": self._get_value(geom, ("phase", "alpha"), "deg"),
+        }
+
+    def afrho(self, geom):
+        """Afρ quanitity given geometrical circumstances.
+
+        Parameters
+        ----------
+        geom: dictionary-like
+            'rh' or 'helio_dist', and 'phase' or 'alpha' angle in au
+            and deg.
+
+        Returns
+        -------
+        afrho: float
+            Afρ parameter in cm.
+
+        """
+        g = self._normalize_geom(geom)
+        afrho = self.afrho1 * g["rh"] ** self.k * self.Phi_c(g["phase"])
+        return afrho
+
+    def mag(self, geom, bandpass, rap=1):
+        """Apparent magnitude within aperture.
+
+        Parameters
+        ----------
+        geom : dictionary-like
+            'rh' or 'helio_dist', 'delta' or 'geo_dist', and 'phase'
+            angle in au and deg.
+
+        bandpass : string
+            LSST bandpass.  One of u, g, r, i, z, y.
+
+        rap : float, optional
+            Aperture radius in arcsec.
+        """
+        g = self._normalize_geom(geom)
+        delta = AU_IN_CM * g["delta"]  # au to cm
+
+        afrho = self.afrho(g)
+        rho = 725e5 * g["delta"] * rap  # arcsec to projected cm
+        dm = -2.5 * np.log10(afrho * rho / (2 * g["rh"] * delta) ** 2)
+        coma = self.m_sun[bandpass] + dm
```

Comment:
should there be better error handling here and check that bandpass is one of LSST bandasses? Otherwise it dies gracefully if someone provided J this would crash I think

Reply:
Correct me if I'm wrong, but if `bandpass` is simply the filter of the observation as passed in from the sorcha code, then this is already handled in sorcha's PPConfigFileParser, which checks that the observing filters supplied by the user in the config file are valid filters from the survey the user has specified.

Reply:
This is now outside of sorcha so someone could edit it.  So given this is an example and someone could send it a filter it does not have and would crash, we should ensure that there is some graceful fail.

Reply:
Understood :)

Reply:
I've updated the code here to check the bandpass against the list of keys available in `self.m_sun`, and if the key is not found, then it will raise a `KeyError` which will be handled by an exception in the `LSSTCometActivity.compute` method, where it will make a call to a new parent class method to log the error both at the terminal and in the error log file.